pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -69,20 +116,38 @@ func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTime
 				}
 			}
 		}()
+
+		go func() {
+			for {
+				select {
+				case <-p.close:
+					return
+				// TODO: this should be 1 minute
+				case <-time.After(2 * time.Second):
+					p.tracker.Write()
+				}
+			}
+		}()
 	}
 
 	return p
 }
 
 // newReaderPool makes a new ReaderPool.
-func newReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics) *ReaderPool {
+func newReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics, userID string) *ReaderPool {
+	lazyReaders := make(map[*LazyBinaryReader]struct{})
 	return &ReaderPool{
 		logger:                logger,
 		metrics:               metrics,
 		lazyReaderEnabled:     lazyReaderEnabled,
 		lazyReaderIdleTimeout: lazyReaderIdleTimeout,
-		lazyReaders:           make(map[*LazyBinaryReader]struct{}),
+		lazyReaders:           lazyReaders,
 		close:                 make(chan struct{}),
+		tracker: HeadersLazyLoadedTracker{
+			lazyReaders: lazyReaders,
+			userID:      userID,
+			persistPath: fmt.Sprintf("/tmp/headers-%s.proto", userID),
```

Comment:
I suppose we can't be writing to `/tmp` because that might not be there next time the pod starts. Do you want to address this in this PR or a followup?

Reply:
Oh yeah, will put that in some data folder pointed by persistence volume. Any suggestion?

Reply:
I think using the sync dir is the easiest way. The [initial sync](https://github.com/grafana/mimir/blob/027536a258762b3f397bc204ceb25a3ab5fbe9a9/pkg/storegateway/bucket.go#L348-L379) is the only place which lists the sync directory. There it ignored anything which isn't a block directory. I think it's safe to use that place to store the files. Maybe a subdirectory (e.g. `loaded`) where there is a file per tenant.